pkg/controller/controller_test.go

```diff
@@ -3444,6 +3597,55 @@ func TestProvisionFromPVC(t *testing.T) {
 			},
 		}
 
+		nilModePV := &v1.PersistentVolume{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: pvNilModeName,
+			},
+			Spec: v1.PersistentVolumeSpec{
+				PersistentVolumeSource: v1.PersistentVolumeSource{
+					CSI: &v1.CSIPersistentVolumeSource{
+						Driver:       driverName,
+						VolumeHandle: "test-nil-volume-id",
+						VolumeAttributes: map[string]string{
+							"storage.kubernetes.io/csiProvisionerIdentity": "test-provisioner",
+						},
+					},
+				},
+				ClaimRef: &v1.ObjectReference{
+					Kind:      "PersistentVolumeClaim",
+					Namespace: srcNamespace,
+					Name:      "nil-mode-claim",
+					UID:       types.UID("fake-nil-claim-uid"),
+				},
+				StorageClassName: fakeSc1,
+			},
+		}
+
+		//clonePVName: "pvSrcBlockMode",
+		blkModePV := &v1.PersistentVolume{
```

Comment:
(nit) `blkModePV` can be instead defined as,
```
                blkModePV := pv.DeepCopy()
		blkModePV.Name = pvSrcBlockMode
		blkModePV.Spec.VolumeMode = &volumeModeBlock
```

And corresponding test cases can be updated to reflect `tc.clonePVName` as `pvSrcBlockMode`

This also removes the need to add `blockModeClaim` below.

Reply:
Actually I noticed some bad behavior in the tests (that I didn't track down).  It seems that we're missing cleanup steps in the tests and the result of trying to do something like you suggest above is things like invalid claims, and unexpected status on PVs.  I'll take a look again, maybe it was something else I had incorrect, or maybe I can more easily identify the problem.  Using debug statements in the catches though, I did notice that our tests don't always fail for the reasons we're expecting them to (sort of a problem), we set `expectErr: true` and let it rip, but it's not always failing for what we expect.

Reply:
I think we should open an issue against the tests, and not necessarily failing for what we expect.  I verified this when trying to copy the pv object into a block object, I suspect we might be doing this in other places.

Reply:
I agree, there were a few flakes and incorrect errors, I have put up this [PR](https://github.com/kubernetes-csi/external-provisioner/pull/411) as an initial cleanup for the tests. It does not ensure the errors are what is expected, when returned are what is expected though.